jax/_src/interpreters/batching.py

```diff
@@ -318,7 +318,7 @@ def flatten_fun_for_vmap(in_tree, *args_flat):
 ### tracer
 
 # TODO(mattjj): use a special sentinel type rather than None
-NotMapped = type(None)
+NotMapped: TypeAlias = type(None)
```

Comment:
I think pyright would still be unhappy about this, unfortunately, because the aliased expression is not a type expression. 

The proper type is either

* `None`, but the runtime does expect `type(None)`, so that doesn't work; or
* `types.NoneType`, but mypy doesn't allow using that in a type annotations (for no good reason imo).

You could add a `# type: ignore` here instead to silence pytype. Not ideal, but likely your best option atm.

---

Repro to confirm my point:

```
from typing import TypeAlias

T: TypeAlias = type(None)

def f(x: T) -> T:
  return None
```

which produces

```
/tmp/foo.py
  /tmp/foo.py:3:16 - error: Invalid expression form for type alias definition (reportInvalidTypeForm)
  /tmp/foo.py:3:16 - error: type() call should not be used in type expression
    Use Type[T] instead (reportInvalidTypeForm)
  /tmp/foo.py:3:16 - error: Call expression not allowed in type expression (reportInvalidTypeForm)
```

Reply:
Hmm, I think I made it fix the issue somehow... I will check again in 16 hours. Thanks for the heads up!

Reply:
Yep, so we just have a misunderstanding. Your illustration is actually the illustration of the fix I care about:

```py
from typing import TypeAlias

T = type(None)

def f(x: T) -> T:
  return None
```

produces:
```
/run/host/tmp/foo.py
  /run/host/tmp/foo.py:4:10 - error: Variable not allowed in type expression (reportInvalidTypeForm)
  /run/host/tmp/foo.py:4:16 - error: Variable not allowed in type expression (reportInvalidTypeForm)
```

while with the ```TypeAlias``` annotation it produces the same thing as in your case:
```
/run/host/tmp/foo.py
  /run/host/tmp/foo.py:3:16 - error: Invalid expression form for type alias definition (reportInvalidTypeForm)
  /run/host/tmp/foo.py:3:16 - error: type() call should not be used in type annotation
    Use Type[T] instead (reportInvalidTypeForm)
  /run/host/tmp/foo.py:3:16 - error: Call expression not allowed in type expression (reportInvalidTypeForm)
```
---
One should note, that the error I care about is the "Variable not allowed in type expression". I am not trying to run pyright on the JAX codebase, I'm running it on the derivative codebase, that just exports stuff from JAX. And, if we split the file into 2, like so (to simulate my scenario):
```py fake_jax.py
from typing import TypeAlias

NotMapped: TypeAlias = type(None)
```

```py main.py
from fake_jax import NotMapped
x: NotMapped
```

Then, when running pyright just on main.py we get no errors! But, without the annotation, we get the "Variable not allowed in type expression". So, the end users (myself included) are happy, but JAX codebase gets a few more pyright errors.

@superbobry WDYT?

Reply:
Ah, I see, you were fixing a bunch of pyright things jax-side in #23453 ... Then probably we could agree on using the type annotation, and then putting ```#pyright: ignore``` in the jax code? like so:
```
# fake_jax.py
from typing import TypeAlias

NotMapped: TypeAlias = type(None) # pyright: ignore
```

Reply:
Let me see if we can make `NotMapped` a proper class instead. I've sketched #23527, checking how many things break as a result :)